### PR TITLE
Add `stream-sum` example

### PR DESCRIPTION
**Pull request by jgraettinger**
Also re-tool `run-consumer` to adopt Viper and ConfigMap for configuration management.

---
This change is [Reviewable](https://reviewable.io/reviews/liveramp/gazette/21)

### CODE REVIEW

**Comment by rupertchen** [comment]
Mostly nits. I think the question about adding a constraint for viper is the most interesting.

---

Reviewed 8 of 8 files at r1, 2 of 2 files at r2, 15 of 15 files at r3.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[.gitignore, line 1 at r1](https://reviewable.io:443/reviews/liveramp/gazette/21#-L30RrHTjqH-wSJcyVEx:-L30RrHTjqH-wSJcyVEy:bsyz9y3) ([raw file](https://github.com/liveramp/gazette/blob/51750a339b08a937b3dc44e2d8a45e104e51f1e1/.gitignore#L1)):*
> ```
> .idea
> ```

While this is common, I prefer keeping IDE-specific ignores out of the project and in an individual's global git ignore, [`$HOME/.config/git/ignore`](https://git-scm.com/docs/gitignore).

---

*[.gitignore, line 2 at r1](https://reviewable.io:443/reviews/liveramp/gazette/21#-L30TSu3PVnhsv_yzyci:-L30TSu3PVnhsv_yzycj:b-zcog62) ([raw file](https://github.com/liveramp/gazette/blob/51750a339b08a937b3dc44e2d8a45e104e51f1e1/.gitignore#L2)):*
> ```
> .idea
> vendor
> ```

Nit: Missing new line.

---

*[build/examples/Dockerfile.stream-sum, line 3 at r3](https://reviewable.io:443/reviews/liveramp/gazette/21#-L30YPoVmLvzbJ_BbA8z:-L30YPoVmLvzbJ_BbA9-:brujoh9) ([raw file](https://github.com/liveramp/gazette/blob/a5805c3c8cde33d44c190207c70006e8953afaf8/build/examples/Dockerfile.stream-sum#L3)):*
> ```
> FROM gazette-build:latest AS build
> 
> COPY examples/stream-sum /go/src/github.com/LiveRamp/gazette/examples/stream-sum
> ```

Nit: Since this path is used several more times,  [`WORKDIR`](https://docs.docker.com/engine/reference/builder/#workdir) could be used to shorten some of the later commands. This file is short and simple enough that it's not an issue either way.

---

*[charts/examples/stream-sum/templates/NOTES.txt, line 0 at r3](https://reviewable.io:443/reviews/liveramp/gazette/21#-L30_a58Kxd-OGAhCJMj:-L30_a58Kxd-OGAhCJMk:bz973et) ([raw file](https://github.com/liveramp/gazette/blob/a5805c3c8cde33d44c190207c70006e8953afaf8/charts/examples/stream-sum/templates/NOTES.txt#L0)):*
Since we have this file, should we add some notes?

---

*[cmd/run-consumer/main.go, line 15 at r1](https://reviewable.io:443/reviews/liveramp/gazette/21#-L30U8aZz-dT0DCvZQBM:-L30U8a_Rt4c4tWCvVwZ:b-9zuev) ([raw file](https://github.com/liveramp/gazette/blob/51750a339b08a937b3dc44e2d8a45e104e51f1e1/cmd/run-consumer/main.go#L15)):*
> ```Go
> 	"github.com/prometheus/client_golang/prometheus"
> 	log "github.com/sirupsen/logrus"
> 	"github.com/spf13/viper"
> ```

A lot of new projects were added to `Gopkg.lock` with the addition of Viper. Should we set a constraint on Viper at least in `Gopkg.toml`?

---

*[examples/stream-sum/topics.go, line 30 at r3](https://reviewable.io:443/reviews/liveramp/gazette/21#-L34xNR2MvNS_hlAy_dg:-L34xNR2MvNS_hlAy_dh:b-kkzsqi) ([raw file](https://github.com/liveramp/gazette/blob/a5805c3c8cde33d44c190207c70006e8953afaf8/examples/stream-sum/topics.go#L30)):*
> ```Go
> 	ID    StreamID // Unique ID of the stream.
> 	SeqNo int      // Monotonic sequence number. One begins a new stream under this unique StreamID.
> 	Data  []byte   // Raw data included in the Value. If empty, this is the streams final chunk.
> ```

Nit: `streams` => `stream's`

---

*[examples/stream-sum/topics.go, line 85 at r3](https://reviewable.io:443/reviews/liveramp/gazette/21#-L34y0RAFwhkWYo1yUzG:-L34y0RAFwhkWYo1yUzH:bxuy0ro) ([raw file](https://github.com/liveramp/gazette/blob/a5805c3c8cde33d44c190207c70006e8953afaf8/examples/stream-sum/topics.go#L85)):*
> ```Go
> }
> 
> // ReadPRNG generates fast but high-quality random entropy into the provided byte slice.
> ```

Nit: `ReadPRNG` => `FillPRNG`.

---

*[examples/stream-sum/chunker/chunker.go, line 12 at r3](https://reviewable.io:443/reviews/liveramp/gazette/21#-L350x0pxBai3ck8P7Sh:-L350x0pxBai3ck8P7Si:brucv5w) ([raw file](https://github.com/liveramp/gazette/blob/a5805c3c8cde33d44c190207c70006e8953afaf8/examples/stream-sum/chunker/chunker.go#L12)):*
> ```Go
> 
> 	"github.com/LiveRamp/gazette/pkg/journal"
> 	"github.com/LiveRamp/gazette/pkg/topic"
> ```

Nit: Import grouping

---

*[examples/stream-sum/summer/summer.go, line 8 at r3](https://reviewable.io:443/reviews/liveramp/gazette/21#-L353scNz6DAsi7gTULy:-L353scNz6DAsi7gTULz:barq6ei) ([raw file](https://github.com/liveramp/gazette/blob/a5805c3c8cde33d44c190207c70006e8953afaf8/examples/stream-sum/summer/summer.go#L8)):*
> ```Go
> 	"encoding/json"
> 
> 	"github.com/LiveRamp/gazette/pkg/journal"
> ```

Nit: Import grouping.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/liveramp/gazette/21#-:-L354fwE6JD2KdGZGzsp:b-dku4ri)*

**Comment by jgraettinger** [comment]
Thanks, PTAL

---

Review status: 7 of 25 files reviewed at latest revision, 9 unresolved discussions.

---

*[build/examples/Dockerfile.stream-sum, line 3 at r3](https://reviewable.io:443/reviews/liveramp/gazette/21#-L30YPoVmLvzbJ_BbA8z:-L3kEYEjTXtp3GS-2Q2H:b-tl9ro6) ([raw file](https://github.com/liveramp/gazette/blob/a5805c3c8cde33d44c190207c70006e8953afaf8/build/examples/Dockerfile.stream-sum#L3)):*
<details><summary><i>Previously, rupertchen (Rupert Chen) wrote…</i></summary><blockquote>

Nit: Since this path is used several more times,  [`WORKDIR`](https://docs.docker.com/engine/reference/builder/#workdir) could be used to shorten some of the later commands. This file is short and simple enough that it's not an issue either way.
</blockquote></details>

`WORKDIR` is `/go/` as defined by the golang base image, no?

I do have a mild preference to keep paths explicit / as they would appear in a plain `go foo`.  A variable indirection means the reader having to resolve what the variable is. This is a bit wordy but blindingly obvious.

---

*[charts/examples/stream-sum/templates/NOTES.txt, line  at r3](https://reviewable.io:443/reviews/liveramp/gazette/21#-L30_a58Kxd-OGAhCJMj:-L3kHH64Dqa3Q10f4PbE:b-896fix) ([raw file](https://github.com/liveramp/gazette/blob/a5805c3c8cde33d44c190207c70006e8953afaf8/charts/examples/stream-sum/templates/NOTES.txt#L)):*
<details><summary><i>Previously, rupertchen (Rupert Chen) wrote…</i></summary><blockquote>

Since we have this file, should we add some notes?
</blockquote></details>

Done.

---

*[cmd/run-consumer/main.go, line 15 at r1](https://reviewable.io:443/reviews/liveramp/gazette/21#-L30U8aZz-dT0DCvZQBM:-L3kHyE1TFO5Bkmodpq5:b4fuz6e) ([raw file](https://github.com/liveramp/gazette/blob/51750a339b08a937b3dc44e2d8a45e104e51f1e1/cmd/run-consumer/main.go#L15)):*
<details><summary><i>Previously, rupertchen (Rupert Chen) wrote…</i></summary><blockquote>

A lot of new projects were added to `Gopkg.lock` with the addition of Viper. Should we set a constraint on Viper at least in `Gopkg.toml`?
</blockquote></details>

Done. My understanding is that dep will permit any version on `ensure` < v2.0.0, right?

---

*[examples/stream-sum/topics.go, line 30 at r3](https://reviewable.io:443/reviews/liveramp/gazette/21#-L34xNR2MvNS_hlAy_dg:-L3kK7An1C3xz6RiaWre:b-896fix) ([raw file](https://github.com/liveramp/gazette/blob/a5805c3c8cde33d44c190207c70006e8953afaf8/examples/stream-sum/topics.go#L30)):*
<details><summary><i>Previously, rupertchen (Rupert Chen) wrote…</i></summary><blockquote>

Nit: `streams` => `stream's`
</blockquote></details>

Done.

---

*[examples/stream-sum/topics.go, line 85 at r3](https://reviewable.io:443/reviews/liveramp/gazette/21#-L34y0RAFwhkWYo1yUzG:-L3kKH_qLpCwAnTX5DeY:b-896fix) ([raw file](https://github.com/liveramp/gazette/blob/a5805c3c8cde33d44c190207c70006e8953afaf8/examples/stream-sum/topics.go#L85)):*
<details><summary><i>Previously, rupertchen (Rupert Chen) wrote…</i></summary><blockquote>

Nit: `ReadPRNG` => `FillPRNG`.
</blockquote></details>

Done.

---

*[examples/stream-sum/chunker/chunker.go, line 12 at r3](https://reviewable.io:443/reviews/liveramp/gazette/21#-L350x0pxBai3ck8P7Sh:-L3kKMpkfflLaxzDCJGA:b-896fix) ([raw file](https://github.com/liveramp/gazette/blob/a5805c3c8cde33d44c190207c70006e8953afaf8/examples/stream-sum/chunker/chunker.go#L12)):*
<details><summary><i>Previously, rupertchen (Rupert Chen) wrote…</i></summary><blockquote>

Nit: Import grouping
</blockquote></details>

Done.

---

*[examples/stream-sum/summer/summer.go, line 8 at r3](https://reviewable.io:443/reviews/liveramp/gazette/21#-L353scNz6DAsi7gTULy:-L3kKnNvzpB97U7j13Lq:b-896fix) ([raw file](https://github.com/liveramp/gazette/blob/a5805c3c8cde33d44c190207c70006e8953afaf8/examples/stream-sum/summer/summer.go#L8)):*
<details><summary><i>Previously, rupertchen (Rupert Chen) wrote…</i></summary><blockquote>

Nit: Import grouping.
</blockquote></details>

Done.

---

*[.gitignore, line 1 at r1](https://reviewable.io:443/reviews/liveramp/gazette/21#-L30RrHTjqH-wSJcyVEx:-L3kDtqvLzvJj9AnTBLc:b-896fix) ([raw file](https://github.com/liveramp/gazette/blob/51750a339b08a937b3dc44e2d8a45e104e51f1e1/.gitignore#L1)):*
<details><summary><i>Previously, rupertchen (Rupert Chen) wrote…</i></summary><blockquote>

While this is common, I prefer keeping IDE-specific ignores out of the project and in an individual's global git ignore, [`$HOME/.config/git/ignore`](https://git-scm.com/docs/gitignore).
</blockquote></details>

Done.

---

*[.gitignore, line 2 at r1](https://reviewable.io:443/reviews/liveramp/gazette/21#-L30TSu3PVnhsv_yzyci:-L3kDzA3zyyQuBDKCb3e:b-896fix) ([raw file](https://github.com/liveramp/gazette/blob/51750a339b08a937b3dc44e2d8a45e104e51f1e1/.gitignore#L2)):*
<details><summary><i>Previously, rupertchen (Rupert Chen) wrote…</i></summary><blockquote>

Nit: Missing new line.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/liveramp/gazette/21#-:-L3kOX71VU3JbrqCf48I:bc8291t)*

**Comment by rupertchen** [comment]
:lgtm:

---

Reviewed 1 of 19 files at r4, 2 of 2 files at r5, 15 of 15 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[cmd/run-consumer/main.go, line 15 at r1](https://reviewable.io:443/reviews/liveramp/gazette/21#-L30U8aZz-dT0DCvZQBM:-L3o9gAV9jou6bEI9yQF:b-85myiv) ([raw file](https://github.com/liveramp/gazette/blob/51750a339b08a937b3dc44e2d8a45e104e51f1e1/cmd/run-consumer/main.go#L15)):*
<details><summary><i>Previously, jgraettinger (Johnny Graettinger) wrote…</i></summary><blockquote>

Done. My understanding is that dep will permit any version on `ensure` < v2.0.0, right?
</blockquote></details>

Correct. Dep is a bit sneaky here. In dep, a [version constraint](https://golang.github.io/dep/docs/Gopkg.toml.html#version-rules) that does not start with an operator automatically uses the `^` operator.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/liveramp/gazette/21#-:-L3oA_PenG3K8wuIcARh:bnfp4nl)*
